File: Front/secondhand-trading-frontend/src/api/product.js
import axios from 'axios'
import { API_CONFIG, getApiUrl } from '../config/api.js'

// 获取图片完整URL的辅助函数
export const getImageUrl = (imageUrl) => {
  if (!imageUrl) {
    console.debug('getImageUrl: empty image URL')
    return ''
  }
  
  try {
    // 如果已经是完整URL则直接返回
    if (imageUrl.startsWith('http') || imageUrl.startsWith('https')) {
      return imageUrl
    }
    
    // 如果是 base64 数据
    if (imageUrl.startsWith('data:image')) {
      return imageUrl
    }    
    // 确保 imageUrl 不以斜杠开头
    const cleanImageUrl = imageUrl.startsWith('/') ? imageUrl.slice(1) : imageUrl
    
    // 返回完整的图片URL
    return `${API_CONFIG.BASE_URL}/${cleanImageUrl}`
    
  } catch (error) {
    console.error('Error in getImageUrl:', error)
    return ''
  }
}

// 规范化商品数据
const normalizeProduct = (product) => {
  return {
    id: product.productId,
    userId: product.userId,
    title: product.title,
    price: product.price,
    description: product.description,
    status: product.status,
    created_at: product.createdAt,
    images: product.images?.map(img => img.imageUrl) || []
  }
}

// 错误处理函数
const handleError = (error) => {
  console.error('API error:', error)
  const response = error.response || {}
  
  // 添加详细的错误日志
  if (error.response) {
    // 服务器响应了，但状态码不在 2xx 范围内
    console.error('Error response:', {
      status: error.response.status,
      data: error.response.data,
      headers: error.response.headers
    })
  } else if (error.request) {
    // 请求已经发出，但没有收到响应
    console.error('No response received:', error.request)
  } else {
    // 在设置请求时发生了错误
    console.error('Request error:', error.message)
  }

  return {
    code: response.status || 500,
    message: response.data?.message || error.message || '请求失败',
    data: null
  }
}

// 创建带基础配置的axios实例
const apiClient = axios.create({
  baseURL: getApiUrl(''),
  headers: {
    'Content-Type': 'application/json',
  },
  withCredentials: true,
  // 添加超时设置
  timeout: 10000,
  // 添加重试配置
  retry: 3,
  retryDelay: 1000
})

// 请求拦截器：添加token
apiClient.interceptors.request.use(config => {
  const token = localStorage.getItem('token')
  if (token) {
    config.headers.Authorization = `Bearer ${token}`
  }
  return config
}, error => {
  return Promise.reject(error)
})

// 添加请求重试拦截器
apiClient.interceptors.response.use(null, async error => {
  const config = error.config;
  
  // 如果配置了重试，且请求失败
  if (config.retry) {
    const backoff = new Promise(resolve => {
      setTimeout(() => {
        resolve();
      }, config.retryDelay || 1000);
    });
    
    // 设置重试次数
    config.retryCount = config.retryCount || 0;
    
    // 如果重试次数小于设置的重试次数
    if (config.retryCount < config.retry) {
      config.retryCount += 1;
      console.log(`Retrying request (${config.retryCount}/${config.retry})`);
      
      // 等待延迟时间后重试
      await backoff;
      return apiClient(config);
    }
  }
  
  return Promise.reject(error);
});

// 原有的响应拦截器
apiClient.interceptors.response.use(
  response => response,
  error => {
    if (error.response?.status === 401) {
      localStorage.removeItem('token')
      localStorage.removeItem('user')
      window.location.href = '/login'
    }
    return Promise.reject(error)
  }
)

export const productApi = {
  // 上传商品（带多图上传）
  addProduct: async (formData) => {
    try {
      const response = await apiClient.post('/products', formData, {
        headers: {
          'Content-Type': 'multipart/form-data'
        }
      })
      return {
        code: response.data.code,
        message: response.data.message,
        data: response.data.data ? normalizeProduct(response.data.data) : null
      }
    } catch (error) {
      return handleError(error)
    }
  },

  // 获取商品列表
  getProducts: async (ids) => {
    try {
      const params = ids ? { ids: ids.join(',') } : {}
      const response = await apiClient.get('/products', { params })
      
      return {
        code: response.data.code,
        message: response.data.message,
        data: response.data.data?.map(normalizeProduct) || []
      }
    } catch (error) {
      return handleError(error)
    }
  },

  // 获取单个商品详情
  getProductDetail: async (id) => {
    try {
      const response = await apiClient.get(`/products/${id}`)
      return {
        code: response.data.code,
        message: response.data.message,
        data: response.data.data ? normalizeProduct(response.data.data) : null
      }
    } catch (error) {
      return handleError(error)
    }
  },

  // 搜索商品
  searchProducts: async (keyword) => {
    try {
      const response = await apiClient.get('/products/search', {
        params: { keyword }
      })
      
      return {
        code: response.data.code,
        message: response.data.message,
        data: {
          products: response.data.data?.map(normalizeProduct) || []
        }
      }
    } catch (error) {
      return handleError(error)
    }
  },

  // 修改商品
  updateProduct: async (id, formData) => {
    try {
      const response = await apiClient.put(`/products/${id}`, formData, {
        headers: {
          'Content-Type': 'multipart/form-data'
        }
      })
      return {
        code: response.data.code,
        message: response.data.message,
        data: response.data.data ? normalizeProduct(response.data.data) : null
      }
    } catch (error) {
      return handleError(error)
    }
  },

  // 删除商品
  deleteProduct: async (id) => {
    try {
      const response = await apiClient.delete(`/products/${id}`, {
        params: { userId: JSON.parse(localStorage.getItem('user'))?.userId }
      })
      return response.data
    } catch (error) {
      return handleError(error)
    }
  }
}

// 获取认证头
const getAuthHeaders = () => {
  const token = localStorage.getItem('token')
  return token ? { 'Authorization': `Bearer ${token}` } : {}
}

File: Back/Python_Basic/app/routes/product.py
import json

from flask import Blueprint, request
from app.services.product_service import ProductService
from app.common.base_response import BaseResponse

product_bp = Blueprint('products', __name__, url_prefix='/api/products')
product_service = ProductService()



def parse_product_dto():
    """从 multipart/form-data 的 files 中解析 productDto 字段"""
    product_dto_file = request.files.get('productDto')
    if not product_dto_file:
        raise ValueError("缺少 productDto 字段")
    try:
        return json.load(product_dto_file)
    except Exception as e:
        raise ValueError(f"productDto 解析失败: {str(e)}")


@product_bp.route('', methods=['POST'])
def add_product():
    try:
        product_data = parse_product_dto()
        images = request.files.getlist('images')
        product = product_service.add_product(product_data, images)
        return BaseResponse.success("商品上传成功", product)
    except ValueError as e:
        return BaseResponse.error(str(e))


@product_bp.route('/<int:id>', methods=['PUT'])
def update_product(id):
    try:
        product_data = parse_product_dto()
        images = request.files.getlist('images')
        product = product_service.update_product(id, product_data, images)
        return BaseResponse.success("商品修改成功", product)
    except ValueError as e:
        return BaseResponse.error(str(e))

@product_bp.route('/<int:id>', methods=['DELETE'])
def delete_product(id):
    user_id = request.args.get('userId', type=int)
    product_service.delete_product(id, user_id)
    return BaseResponse.success("商品删除成功")


@product_bp.route('/<int:id>', methods=['GET'])
def get_product(id):
    product = product_service.get_product_by_id(id)
    return BaseResponse.success("商品获取成功", product)


@product_bp.route('', methods=['GET'])
def get_all_products():
    products = product_service.get_all_products()
    return BaseResponse.success("商品列表获取成功", products)


@product_bp.route('/user', methods=['GET'])
def get_user_products():
    user_id = request.args.get('userId', type=int)
    products = product_service.get_user_products(user_id)
    return BaseResponse.success("用户商品列表获取成功", products)


@product_bp.route('/search', methods=['GET'])
def search_products():
    keyword = request.args.get('keyword', '')
    products = product_service.search_products(keyword)
    return BaseResponse.success("搜索成功", products)

